documentation/adr/0021-errors-and-warnings.md

```diff
@@ -0,0 +1,104 @@
+  ```matlab
+  function [data_out, fitdata, ok, mess] = fit (obj, varargin)
+
+  if nargout<3
+      throw_error = true;
+  else
+      throw_error = false;
+  end
+
+  [...]
+
+  if ~ok_sim
+    ok = false;
+    if throw_error, error_message(mess), else, return, end
+  end
+
+  (@mfclass/fit)
+  ```
+- Throwing MATLAB errors and performing global catches and rethrows
+  ```matlab
+  try
+     obj = set_bfun@mfclass (obj, varargin{:});
+  catch ME
+      error(ME.message)
+  end
+
+  (@mfclass_tobyfit/set_bfun)
+  ```
+
+### Issue
+
+There is inconsistency throughout PACE about how errors and warnings are handled,
+which leads to maintainability issues and confusing code as well as a decision
+on the developer's part about which path to follow.
+
+The fact that errors are not reported from where they are triggered means that the backtrace
+will also report the wrong location, rendering the backtrace less useful for the developer.
+It is argued that this is in order to not deter new users who will see long tracebacks, however,
+the counter argument to this is that a user should be worried if they are doing erroneous operations
+and should understand that the traceback is intended for the developer rather than them.
+
+## Alternatives
+
+There are situations in which the developer may wish to not have MATLAB explicitly error out when
+a user, for example, simply inputs bad arguments. In these situations a mechanism exists for handling
+known or expected exception classes.
+
+```matlab
+try
+   ...
+   error('PACE:errorClass', errorMessage)
+catch ME
+   switch ME.identifier
+   case 'PACE:errorClass'
+     warning("Error identified and caught")
+   case {'PACE:errorClass1', 'PACE:errorClass2'}
+     warning("Error is class1 or 2")
+   otherwise
+     rethrow(ME)
+   end
+end
+```
+**N.B.** `rethrow` will throw the original error with the original backtrace meaning that
+the developer can still determine where the problem occurred. The warning, however, will trigger with a
+backtrace pointing to its location, thereby known/expected exceptions will not deter the average user.
+
+In MATLAB there are also ways to modify existing warning messages to add clarification:
+```matlab
+try
+   C = [A; B];
+catch ME
+   switch ME.identifier
+   case 'MATLAB:catenate:dimensionMismatch'
+      msg = ['Dimension mismatch occurred: First argument has ', ...
+            num2str(size(A,2)),' columns while second has ', ...
+            num2str(size(B,2)),' columns.'];
+        causeException = MException('MATLAB:myCode:dimensions',msg);
+        ME = addCause(ME,causeException);
+   end
+   rethrow(ME)
+end
+```
+
+## Recommendations
+
+- Replace all handling via args (`ok`, `mess`  syntax) with try-throw-catch process
+- Use `rethrow` over `throw(ME.message)` which
+  1. Destroys the identifier
+  2. Obscures the backtrace
+- Never perform a global catch. Specific catches should be used where relevant, preferably using `switch` for clarity.
```

Comment:
There's another (non-default) MLint setting we could enable for this 
![image](https://user-images.githubusercontent.com/46960113/107774418-df7b6680-6d36-11eb-91bb-329f70a17b82.png)

Reply:
The issue here is that I don't think MATLAB's catch will check for type (as in python)
```python
try:
   raise NameError()
except NameError:
except TypeError:
...
```
I suspect the linter will know whether the subsequent catch checks the identifier. I suspect it's only as smart as "Oh, `catch ME`", that's fine, then. 

I spent some time trying to test this, but the code analyser just shows a blank web browser for whatever reason.

Reply:
I think you're right, but at least we'll get warnings for a bare `try-catch`. If there is a situation where we capture the error, but don't do anything with it, hopefully the "unused variable" warnings can alert us that something's wrong.

Reply:
True, but there are cases of:
```
try
   ...
catch ME
   warning(ME.message)
end
```
I suppose regardless we take all the help we can get.